arrows/vxl/high_pass_filter.cxx

```diff
@@ -0,0 +1,489 @@
+      {
+        const PixType& val = grey( i, j );
+        const PixType& avg1 = smoothed( i - offset, j );
+        const PixType& avg2 = smoothed( i + offset, j );
+
+        diff1 = ( val > avg1 ? val - avg1 : avg1 - val );
+        diff2 = ( val > avg2 ? val - avg2 : avg2 - val );
+
+        output( i, j ) = std::min( diff1, diff2 );
+      }
+    }
+  }
+
+  // Given an input grayscale image, and a smoothed version of this greyscale
+  // image, calculate the bidirectional filter response in the vertical
+  // direction.
+  template < typename PixType >
+  void
+  vertical_box_bidirectional_pass( const vil_image_view< PixType >& grey,
+                                   const vil_image_view< PixType >& smoothed,
+                                   vil_image_view< PixType >& output,
+                                   unsigned kernel_height )
+  {
+    vil_image_view< PixType > grey_t = vil_transpose( grey );
+    vil_image_view< PixType > smoothed_t = vil_transpose( smoothed );
+    vil_image_view< PixType > output_t = vil_transpose( output );
+
+    horizontal_box_bidirectional_pass( grey_t,
+                                       smoothed_t,
+                                       output_t,
+                                       kernel_height );
+  }
+
+  // Perform bidirectional filtering
+  template < typename PixType >
+  vil_image_view< PixType >
+  bidirection_box_filter( const vil_image_view< PixType >& grey_img )
+  {
+    vil_image_view< PixType > output;
+    output.set_size( grey_img.ni(), grey_img.nj(), 3 );
+
+    // Create views of each plane
+    vil_image_view< PixType > filter_x = vil_plane( output, 0 );
+    vil_image_view< PixType > filter_y = vil_plane( output, 1 );
+    vil_image_view< PixType > filter_xy = vil_plane( output, 2 );
+
+    // Report the difference between the pixel value, and all of the smoothed
+    // responses. The xy channel is used as a temporary buffer to avoid
+    // additional memory allocation
+    box_average_vertical( grey_img, filter_xy, kernel_height );
+    horizontal_box_bidirectional_pass( grey_img, filter_xy, filter_x,
+                                       kernel_width );
+    box_average_horizontal( grey_img, filter_xy, kernel_width );
+    vertical_box_bidirectional_pass( grey_img, filter_xy, filter_y,
+                                     kernel_height );
+    vil_math_image_max( filter_x, filter_y, filter_xy );
+
+    return output;
+  }
+
+  // Function which returns an image view of the even rows of an image
+  template < typename PixType >
+  inline vil_image_view< PixType >
+  even_rows( const vil_image_view< PixType >& im )
+  {
+    return vil_image_view< PixType >( im.memory_chunk(), im.top_left_ptr(),
+                                      im.ni(), ( im.nj() + 1 ) / 2,
+                                      im.nplanes(),
+                                      im.istep(), im.jstep() * 2,
+                                      im.planestep() );
```

Comment:
```suggestion
    return vil_image_view< PixType >{
      im.memory_chunk(), im.top_left_ptr(),
      im.ni(), ( im.nj() + 1 ) / 2, im.nplanes(),
      im.istep(), im.jstep() * 2, im.planestep() };
```

As previously noted, up to you where you prefer to put the line breaks; this one is very much a "have you considered?". Using `{}`s would be nice, though :slightly_smiling_face:. (Same thing a few lines below.)